File: software/examples/javascript/ExampleCallback.js
var Tinkerforge = require('tinkerforge');

var HOST = 'localhost';
var PORT = 4223;
var UID = 'XYZ'; // Change XYZ to the UID of your Performance DC Bricklet

var ipcon = new Tinkerforge.IPConnection(); // Create IP connection
var pdc = new Tinkerforge.BrickletPerformanceDC(UID, ipcon); // Create device object

ipcon.connect(HOST, PORT,
    function (error) {
        console.log('Error: ' + error);
    }
); // Connect to brickd
// Don't use device before ipcon is connected

ipcon.on(Tinkerforge.IPConnection.CALLBACK_CONNECTED,
    function (connectReason) {
        // Enable velocity reached callback
        pdc.setVelocityReachedCallbackConfiguration(true);

        // The acceleration has to be smaller or equal to the maximum
        // acceleration of the DC motor, otherwise the velocity reached
        // callback will be called too early
        pdc.setMotion(4096, 4096); // Slow acceleration (12.5 %/s)
        pdc.setVelocity(32767); // Full speed forward (100 %)

        // Enable motor power
        pdc.setEnabled(true);
    }
);

// Register velocity reached callback
pdc.on(Tinkerforge.BrickletPerformanceDC.CALLBACK_VELOCITY_REACHED,
    // Use velocity reached callback to swing back and forth
    // between full speed forward and full speed backward
    function (velocity) {
        if(velocity == 32767) {
            console.log('Velocity: Full speed forward, now turning backward');
            pdc.setVelocity(-32767);
        }
        else if(velocity === -32767) {
            console.log('Velocity: Full speed backward, now turning forward');
            pdc.setVelocity(32767);
        }
        else {
            console.log('Error'); // Can only happen if another program sets velocity
        }
    }
);

console.log('Press key to exit');
process.stdin.on('data',
    function (data) {
        // Stop motor before disabling motor power
        pdc.setMotion(4096, 16384); // Fast decceleration (50 %/s) for stopping
        pdc.setVelocity(0); // Request motor stop

        setTimeout(function () {
            pdc.setEnabled(false); // Disable motor power

            ipcon.disconnect();
            process.exit(0);
        }, 2000); // Wait for motor to actually stop: velocity (100 %) / decceleration (50 %/s) = 2 s
    }
);

File: software/examples/perl/example_callback.pl
#!/usr/bin/perl

use strict;
use Tinkerforge::IPConnection;
use Tinkerforge::BrickletPerformanceDC;

use constant HOST => 'localhost';
use constant PORT => 4223;
use constant UID => 'XYZ'; # Change XYZ to the UID of your Performance DC Bricklet

my $ipcon = Tinkerforge::IPConnection->new(); # Create IP connection
our $pdc = Tinkerforge::BrickletPerformanceDC->new(&UID, $ipcon); # Create device object

# Use velocity reached callback to swing back and forth
# between full speed forward and full speed backward
sub cb_velocity_reached
{
    my ($velocity) = @_;

    if ($velocity == 32767)
    {
        print "Velocity: Full speed forward, now turning backward\n";
        $pdc->set_velocity(-32767);
    }
    elsif ($velocity == -32767)
    {
        print "Velocity: Full speed backward, now turning forward\n";
        $pdc->set_velocity(32767);
    }
    else
    {
        print "Error\n"; # Can only happen if another program sets velocity
    }
}

$ipcon->connect(&HOST, &PORT); # Connect to brickd
# Don't use device before ipcon is connected

# Register velocity reached callback to subroutine cb_velocity_reached
$pdc->register_callback($pdc->CALLBACK_VELOCITY_REACHED, 'cb_velocity_reached');

# Enable velocity reached callback
$pdc->set_velocity_reached_callback_configuration(1);

# The acceleration has to be smaller or equal to the maximum
# acceleration of the DC motor, otherwise the velocity reached
# callback will be called too early
$pdc->set_motion(4096, 4096); # Slow acceleration (12.5 %/s)
$pdc->set_velocity(32767); # Full speed forward (100 %)

# Enable motor power
$pdc->set_enabled(1);

print "Press key to exit\n";
<STDIN>;

# Stop motor before disabling motor power
$pdc->set_motion(4096, 16384); # Fast decceleration (50 %/s) for stopping
$pdc->set_velocity(0); # Request motor stop
sleep(2); # Wait for motor to actually stop: velocity (100 %) / decceleration (50 %/s) = 2 s
$pdc->set_enabled(0); # Disable motor power

$ipcon->disconnect();
